### PR TITLE
Report bespoke error for restricted types parsed as type arguments

### PR DESCRIPTION
**Pull request by dsainati1**
See discussion in https://discord.com/channels/613813861610684416/1108479699732152503/1186788225365327952

## Description

Report a better error when legacy restricted types are encountered as type arguments
______

- [X] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [X] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

**Review by SupunS — APPROVED** [review]

File: runtime/parser/expression.go [hunk 1]
```diff
@@ -601,7 +601,7 @@ func defineLessThanOrTypeArgumentsExpression() {
 				return nil
 			}()
 
-			// `err` is nil means the expression is an invocation
+			// `err` is nil means the expression is an invocation,
```
Comment (turbolent) [inline runtime/parser/expression.go:604]: Typo?
```suggestion
			// `err` is nil means the expression is an invocation
```

File: runtime/parser/expression.go [hunk 2]
```diff
@@ -640,6 +640,12 @@ func defineLessThanOrTypeArgumentsExpression() {
 
 			} else {
 
+				// if the error specifically occurred during parsing restricted types,
+				// this is still an invocation and we should report that error instead
+				if strings.Contains(err.Error(), "restricted type") {
```
Comment (turbolent) [inline runtime/parser/expression.go:645]: Using the error message to detect this case is brittle: if the error message is changed, the code here will still compile, but will fail at run-time.

Instead of returning a generic `SyntaxError` for the restricted type, introduce and report a new error type, and check for it here.

File: runtime/parser/parser_test.go [hunk 3]
```diff
@@ -506,7 +506,7 @@ func TestParseBuffering(t *testing.T) {
 
           fun isneg(x: SignedFixedPoint): Bool {   /* I kinda forget what this is all about */
               return x                             /* but we probably need to figure it out */
-                     <                             /* ************/((TODO?{/*))************ *//
+                     <                             /* ************/((TODO?/*))************ *//
```
Comment (turbolent) [inline runtime/parser/parser_test.go:509]: Why was this change necessary?

**Review by turbolent — COMMENTED** [review]
Nice! Thank you for figuring out what's wrong and fixing it

File: runtime/parser/parser_test.go (hunk 3 above)
Comment (dsainati1) [reply runtime/parser/parser_test.go:509]: It's not strictly necessary, but the presence of the `{` here makes the parser think that the reason this failed is because it was trying to parse a restricted type. The test immediately after this keeps the `{` and changes its error message to be the restricted type parsing failure message, so here I chose to remove the `{` to test both behaviors.

File: runtime/parser/errors.go [hunk 4]
```diff
@@ -318,3 +318,28 @@ func (e *CustomDestructorError) Error() string {
 func (e *CustomDestructorError) SecondaryError() string {
 	return "remove the destructor definition"
 }
+
+// RestrictedTypeError
+
+type RestrictedTypeError struct {
+	Pos ast.Position
```
Comment (turbolent) [inline runtime/parser/errors.go:325]: This works, but providing a range would be nicer

**Review by turbolent — APPROVED** [review]
Looks good! Maybe consider replacing the start position with a whole range, given it is available
